lineoperations/src/linefunctions.c

```diff
@@ -22,302 +23,237 @@
-		lines[i] = sci_get_line(doc->editor->sci, i);
+gint
+rmdupst(gchar **lines, gint num_lines, gchar *new_file)
+{
+	gchar *nf_end  = new_file;     /* points to last char of new_file */
+	gchar *lineptr = (gchar *)" "; /* temporary line pointer */
+	gint  i        = 0;            /* iterator */
+	gint  changed  = 0;            /* number of lines removed */
 
 	/* sort **lines ascending */
-	qsort(lines, total_num_lines, sizeof(gchar *), compare_asc);
+	qsort(lines, num_lines, sizeof(gchar *), compare_asc);
 
 	/* loop through **lines, join first occurances into one str (new_file) */
-	for(i = 0; i < total_num_lines; i++)
+	for(i = 0; i < num_lines; i++)
+	{
 		if(strcmp(lines[i], lineptr) != 0)
 		{
-			lineptr  = lines[i];
-			nf_end   = g_stpcpy(nf_end, lines[i]);
+			changed++;     /* number of lines kept */
+			lineptr = lines[i];
+			nf_end  = g_stpcpy(nf_end, lines[i]);
 		}
+	}
 
-	/* set new document */
-	sci_set_text(doc->editor->sci, new_file);
-
-	/* free used memory */
-	for(i = 0; i < total_num_lines; i++)
-		g_free(lines[i]);
-	g_free(lines);
-	g_free(new_file);
+	/* return the number of lines deleted */
+	return -(num_lines - changed);
 }
 
 
 /* Remove Duplicate Lines, ordered */
-void rmdupln(GeanyDocument *doc) {
-	gint  total_num_chars;  /* number of characters in the document */
-	gint  total_num_lines;  /* number of lines in the document */
-	gchar **lines;          /* array to hold all lines in the document */
-	gchar *new_file;        /* *final* string to replace current document */
-	gchar *nf_end;          /* points to end of new_file */
-	gint  i;                /* iterator */
-	gint  j;                /* iterator */
-	gboolean *to_remove;    /* flag to 'mark' which lines to remove */
-
-	total_num_chars = sci_get_length(doc->editor->sci);
-	total_num_lines = sci_get_line_count(doc->editor->sci);
-	lines           = g_malloc(sizeof(gchar *) * total_num_lines);
-	new_file        = g_malloc(sizeof(gchar) * (total_num_chars+1));
-	new_file[0]     = '\0';
-	nf_end          = new_file;
-
-
-	/* if file is not empty, ensure that the file ends with newline */
-	if(total_num_lines != 1)
-		ensure_final_newline(doc->editor, total_num_lines);
-
-	/* copy *all* lines into **lines array */
-	for(i = 0; i < total_num_lines; i++)
-		lines[i] = sci_get_line(doc->editor->sci, i);
+gint
+rmdupln(gchar **lines, gint num_lines, gchar *new_file)
+{
+	gchar *nf_end  = new_file;  /* points to last char of new_file */
+	gint  i        = 0;         /* iterator */
+	gint  j        = 0;         /* iterator */
+	gboolean *to_remove = NULL; /* flag to 'mark' which lines to remove */
+	gint  changed  = 0;         /* number of lines removed */
+
 
 	/* allocate and set *to_remove to all FALSE
 	 * to_remove[i] represents whether lines[i] should be removed  */
-	to_remove = g_malloc(sizeof(gboolean) * total_num_lines);
-	for(i = 0; i < (total_num_lines); i++)
+	to_remove = g_malloc(sizeof(gboolean) * num_lines);
+	for(i = 0; i < (num_lines); i++)
```

Comment:
Same as before: You should check whether num_lines > 0  even you might only call it from your code